src/core/server/saved_objects/deprecations/unknown_object_types.ts

```diff
@@ -0,0 +1,168 @@
+  kibanaVersion,
+  kibanaConfig,
+  savedObjectsConfig,
+}: {
+  types: string[];
+  typeRegistry: ISavedObjectTypeRegistry;
+  savedObjectsConfig: SavedObjectConfig;
+  kibanaConfig: KibanaConfigType;
+  kibanaVersion: string;
+}) => {
+  return [
+    ...new Set(
+      types.map((type) =>
+        getIndexForType({
+          type,
+          typeRegistry,
+          migV2Enabled: savedObjectsConfig.migration.enableV2,
+          kibanaVersion,
+          defaultIndex: kibanaConfig.index,
+        })
+      )
+    ),
+  ];
+};
+
+const getUnknownTypesQuery = (knownTypes: string[]): estypes.QueryDslQueryContainer => {
+  return {
+    bool: {
+      must_not: knownTypes.map((type) => ({
+        term: { type },
+      })),
+    },
+  };
+};
+
+const getUnknownSavedObjects = async ({
+  typeRegistry,
+  esClient,
+  kibanaConfig,
+  savedObjectsConfig,
+  kibanaVersion,
+}: UnknownTypesDeprecationOptions) => {
+  const knownTypes = getKnownTypes(typeRegistry);
+  const targetIndices = getTargetIndices({
+    types: knownTypes,
+    typeRegistry,
+    kibanaConfig,
+    kibanaVersion,
+    savedObjectsConfig,
+  });
+  const query = getUnknownTypesQuery(knownTypes);
+
+  const { body } = await esClient.asInternalUser.search<SavedObjectsRawDocSource>({
+    index: targetIndices,
+    body: {
+      size: 10000,
+      query,
+    },
+  });
+  const { hits: unknownDocs } = body.hits;
+
+  return unknownDocs.map((doc) => ({ id: doc._id, type: doc._source?.type ?? 'unknown' }));
+};
+
+export const getUnknownTypesDeprecations = async (
+  options: UnknownTypesDeprecationOptions
+): Promise<DeprecationsDetails[]> => {
+  const deprecations: DeprecationsDetails[] = [];
+  const unknownDocs = await getUnknownSavedObjects(options);
+  if (unknownDocs.length) {
+    deprecations.push({
+      title: i18n.translate('core.savedObjects.deprecations.unknownTypes.title', {
+        defaultMessage: 'Saved objects with unknown types are present in Kibana system indices',
+      }),
+      message: i18n.translate('core.savedObjects.deprecations.unknownTypes.message', {
+        defaultMessage:
+          'Upgrades will fail for 8.0+ because documents were found for unknown saved object types.' +
+          `To ensure that upgrades will succeed in the future, either re-enable plugins or delete these documents from the Kibana indices`,
+      }),
+      level: 'critical',
+      requireRestart: false,
+      deprecationType: undefined, // not config nor feature...
+      correctiveActions: {
+        manualSteps: [
+          i18n.translate('core.savedObjects.deprecations.unknownTypes.manualSteps.1', {
+            defaultMessage: 'If plugins are disabled, re-enable the, then restart Kibana.',
+          }),
+          i18n.translate('core.savedObjects.deprecations.unknownTypes.manualSteps.2', {
+            defaultMessage:
+              'If no plugins are disabled, or if enabling them does not fix the issue, delete the documents.',
+          }),
```

Comment:
Not really sure what to put in the `manualSteps`, as we're mostly expecting the user to use the automated resolve instead.

We could potentially list all the faulty objects as we're doing in the `checkForUnknownDocs` migration task's warning, but that really don't work well in the UA UI.

Reply:
The steps make sense to me. Maybe we can specify how to find these document ids by checking the server logs.